File: code/go/2021/apr/triangle.go
package apr

func minimumTotal(tri [][]int) int {
	if tri == nil {
		return 0
	}

	for row := len(tri) - 2; row >= 0; row-- {
		for col := 0; col < len(tri[row]); col++ {
			tri[row][col] += min(tri[row+1][col], tri[row+1][col+1])
		}
	}
	return tri[0][0]
}

func min(a, b int) int {
	if a > b {
		return b
	}
	return a
}

File: archive/java/src/main/java/cn/az/code/year2021/apr/Triangle.java
package cn.az.code.year2021.apr;

import java.util.List;

/**
 * @author ycpang
 * @since 2021-04-21 15:26
 */
public class Triangle {

    public int minimumTotal(List<List<Integer>> triangle) {
        for (int i = triangle.size() - 2; i >= 0; i--) {
            for (int j = 0; j < triangle.get(i).size(); j++) {
                triangle.get(i).set(j, triangle.get(i).get(j) + Math.min(triangle.get(i + 1).get(j), triangle.get(i + 1).get(j + 1)));
            }
        }
        return triangle.get(0).get(0);
    }
}
